File: flask_app/app/dao/product_dao.py
from app.models.product import Product
from app.models import db


class ProductDAO:

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        query = Product.query
        filters = filters or {}

        if filters.get("category_id"):
            query = query.filter(Product.category_id == filters["category_id"])

        if filters.get("discounted") is not None:
            if filters["discounted"]:
                query = query.filter(Product.discount_id.isnot(None))
            else:
                query = query.filter(Product.discount_id.is_(None))

        if filters.get("in_stock") is not None:
            if filters["in_stock"]:
                query = query.filter(Product.stock_quantity > 0)
            else:
                query = query.filter(Product.stock_quantity <= 0)

        if filters.get("exclusive") is not None:
            query = query.filter(
                Product.is_membership_exclusive == filters["exclusive"]
            )

        if filters.get("product"):
            name_filter = filters["product"]
            query = query.filter(Product.name.ilike(f"%{name_filter}%"))

        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        return pagination.items, pagination.total

    @staticmethod
    def get_by_id(product_id):
        return Product.query.filter_by(id=product_id).first()

    @staticmethod
    def create(data):
        product = Product(**data)
        db.session.add(product)
        db.session.commit()
        return product

    @staticmethod
    def update(product_id, data):
        product = ProductDAO.get_by_id(product_id)
        if not product:
            return None
        for key, value in data.items():
            if hasattr(product, key):
                setattr(product, key, value)
        db.session.commit()
        return product

    @staticmethod
    def commit(product):
        db.session.add(product)
        db.session.commit()

    @staticmethod
    def soft_delete(product_id):
        # Assuming product does not have is_active, so we actually delete or mark in some way
        product = ProductDAO.get_by_id(product_id)
        if not product:
            return False
        db.session.delete(product)
        db.session.commit()
        return True

File: springboot_api/src/main/java/com/example/membership_api/dao/ProductDAO.java
package com.example.membership_api.dao;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Repository;

import com.example.membership_api.model.Product;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.TypedQuery;

@Repository
public class ProductDAO {

    @PersistenceContext
    private EntityManager em;

    public List<Product> getAll(int page, int perPage, Map<String, Object> filters) {
        StringBuilder queryBuilder = new StringBuilder("FROM Product p WHERE 1=1 ");

        Map<String, Object> params = new HashMap<>();

        if (filters != null) {
            if (filters.containsKey("category_id")) {
                queryBuilder.append("AND p.categoryId = :categoryId ");
                params.put("categoryId", filters.get("category_id"));
            }
            if (filters.containsKey("discounted")) {
                if (Boolean.TRUE.equals(filters.get("discounted"))) {
                    queryBuilder.append("AND p.discountId IS NOT NULL ");
                } else {
                    queryBuilder.append("AND p.discountId IS NULL ");
                }
            }
            if (filters.containsKey("in_stock")) {
                if (Boolean.TRUE.equals(filters.get("in_stock"))) {
                    queryBuilder.append("AND p.stockQuantity > 0 ");
                } else {
                    queryBuilder.append("AND p.stockQuantity <= 0 ");
                }
            }
            if (filters.containsKey("exclusive")) {
                queryBuilder.append("AND p.isMembershipExclusive = :exclusive ");
                params.put("exclusive", filters.get("exclusive"));
            }
            if (filters.containsKey("product")) {
                queryBuilder.append("AND LOWER(p.name) LIKE :product ");
                params.put("product", "%" + filters.get("product").toString().toLowerCase() + "%");
            }
        }

        TypedQuery<Product> query = em.createQuery(queryBuilder.toString(), Product.class);

        params.forEach(query::setParameter);

        query.setFirstResult((page - 1) * perPage);
        query.setMaxResults(perPage);

        return query.getResultList();
    }

    public long getTotalCount(Map<String, Object> filters) {
        StringBuilder queryBuilder = new StringBuilder("SELECT COUNT(p) FROM Product p WHERE 1=1 ");

        Map<String, Object> params = new HashMap<>();

        if (filters != null) {
            if (filters.containsKey("category_id")) {
                queryBuilder.append("AND p.categoryId = :categoryId ");
                params.put("categoryId", filters.get("category_id"));
            }
            if (filters.containsKey("discounted")) {
                if (Boolean.TRUE.equals(filters.get("discounted"))) {
                    queryBuilder.append("AND p.discountId IS NOT NULL ");
                } else {
                    queryBuilder.append("AND p.discountId IS NULL ");
                }
            }
            if (filters.containsKey("in_stock")) {
                if (Boolean.TRUE.equals(filters.get("in_stock"))) {
                    queryBuilder.append("AND p.stockQuantity > 0 ");
                } else {
                    queryBuilder.append("AND p.stockQuantity <= 0 ");
                }
            }
            if (filters.containsKey("exclusive")) {
                queryBuilder.append("AND p.isMembershipExclusive = :exclusive ");
                params.put("exclusive", filters.get("exclusive"));
            }
            if (filters.containsKey("product")) {
                queryBuilder.append("AND LOWER(p.name) LIKE :product ");
                params.put("product", "%" + filters.get("product").toString().toLowerCase() + "%");
            }
        }

        TypedQuery<Long> query = em.createQuery(queryBuilder.toString(), Long.class);
        params.forEach(query::setParameter);

        return query.getSingleResult();
    }

    public Product getById(String productId) {
        return em.find(Product.class, productId);
    }

    public Product create(Product product) {
        em.persist(product);
        return product;
    }

    public Product update(Product existing, Map<String, Object> updates) {
        if (updates.containsKey("name")) existing.setName((String) updates.get("name"));
        if (updates.containsKey("description")) existing.setDescription((String) updates.get("description"));
        if (updates.containsKey("category_id")) existing.setCategoryId((String) updates.get("category_id"));
        if (updates.containsKey("stock_quantity")) existing.setStockQuantity((Integer) updates.get("stock_quantity"));
        if (updates.containsKey("discount_id")) existing.setDiscountId((String) updates.get("discount_id"));
        if (updates.containsKey("price")) existing.setPrice(Double.valueOf(updates.get("price").toString()));
        if (updates.containsKey("is_membership_exclusive"))
            existing.setIsMembershipExclusive((Boolean) updates.get("is_membership_exclusive"));
        if (updates.containsKey("min_tier_required"))
            existing.setMinTierRequired((Integer) updates.get("min_tier_required"));
        return existing;
    }

    public void commit(Product product) {
        em.merge(product);
    }

    public boolean softDelete(String productId) {
        Product product = getById(productId);
        if (product == null) return false;
        em.remove(product);
        return true;
    }
}
